### DISCUSSION TITLE
How to include arguments with key for fallback data in <SWRConfig> provider?

### DISCUSSION THREAD

**Discussion by tylermassey** [vercel/swr#1447] [Q&A] [answered]
The examples for providing initial data via the `fallback` option in the `SWRConfig` provider make sense.

```ts
<SWRConfig value={{
  fallback: {
    '/api/user': { name: 'Bob', ... },
    '/api/items': ...,
    ...
  }
}}>
```

But what if a request needs arguments such as, for example, when using a GraphQL API? 

```ts
const { data } = useSWR([ GRAPHQL_ENDPOINT, graphqlQuery ], fetcher);
```

```ts
fallback: {
  '/api/graphql': { data }
}
```

The above does not seem sufficient when a GraphQL query is used as part of the cache key.

**Comment by shuding** [comment, marked as answer]
```js
import { unstable_serialize } from 'swr'

…

fallback: {
  [unstable_serialize([GRAPHQL_ENDPOINT, graphqlQuery])]: data
}
```

You have to serialize the arguments like this, since it’s not simply a string anymore. We will have this documented soon! Thanks for the feedback.

**Reply by tylermassey** [reply]
Thanks for the help!

**Comment by joggienl** [comment]
@shuding I've got a couple of questions: 

- for what version of SWR does this work?
- will this always be named `unstable_serialize`? (kinda "vague")

I have a NextJS 11 project with SWR 1.1.1 and I cannot get this to work.

I've setup a simple test page that queries a pokemon api with two arguments and just displays a list with all the moves this pokemon (pikachu) has. Somehow it does not work. Perhaps I am doing something wrong. I also took a look at which I guess is the actual documentation for this feature.

I tried to put the example in a codesanbox, but for some reason that kept crashing.

```js
import PropTypes from 'prop-types';
import useSWR, { SWRConfig, unstable_serialize } from 'swr';

const API = 'https://pokeapi.co/api/v2';
const fetcher = (what, who) =>
  fetch(`${API}/${what}/${who}`, { mode: 'cors' }).then((res) => res.json());

function Pokemon() {
  const { data } = useSWR(['pokemon', 'pikachu']);

  // data should never be undefined!
  console.log(!!data);

  return (
    <>
      <h2>pikachu moves:</h2>
      <ul>
        {data.moves.map((move, index) => (
          <li key={`move-${index}`}>{move.move.name}</li>
        ))}
      </ul>
    </>
  );
}

export default function PokemonPage({ fallback }) {
  return (
    <SWRConfig value={{ fallback, fetcher }}>
      <div>
        <Pokemon />
      </div>
    </SWRConfig>
  );
}

PokemonPage.propTypes = {
  fallback: PropTypes.shape({}).isRequired,
};

export async function getStaticProps() {
  const pikachu = await fetcher('pokemon', 'pikachu');

  return {
    props: {
      fallback: {
        [unstable_serialize(['pokemon', 'pikachu'])]: pikachu,
      },
    },
  };
}
```

**Reply by joggienl** [reply]
oh my... I am going nuts over this feature haha.

I tried to apply my example in StackBlitz to share a working non-working example. To my surprise it worked exactly as described...

This should work in any Chrome browser: https://stackblitz.com/edit/nextjs-lxla6c?devtoolsheight=33&file=pages/index.js

I restarted my local nextjs project and now it works there as well. Well, this is kinda magic 😅 .

I did notice that the cache-ids in the situation where it did not work started with `$args...`.. In this working situation they start with `@/pokemon/pikachu`..

Weird.

Stil my question about the name of this function holds!

**Reply by shuding** [reply]
I'll take a look tomorrow! This API should work since 1.0.0, and we will likely keep the name until we figure out a better API (so you don't need to always serialize the keys like that, or at least make it easier).
